src/main/java/org/junit/runners/RuleContainer.java

```diff
@@ -0,0 +1,136 @@
+package org.junit.runners;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.IdentityHashMap;
+import java.util.List;
+
+import org.junit.rules.MethodRule;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.FrameworkMethod;
+import org.junit.runners.model.Statement;
+
+/**
+ * @since 4.13
+ */
+public final class RuleContainer {
```

Comment:
The only question is how likely someone might want to reuse it? The only part which should stay "hidden" here is the `ThreadLocal` field, the rest of the class can be moved into the `internal` package.